Desugar simple UCS and add old tests to the staging area

chengluyu opened the pull request:
This PR adds `Desugarer` to hkmc2. It supports simple UCS expressions now.

chengluyu committed:
Port core abstract syntax and fix type errors (incorrect though)

chengluyu committed:
Merge branch 'hkmc2-bbml' from NeilKleistGao into hkmc2-ucs

chengluyu committed:
Desugar simple UCS and add old tests

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/bbml/TypeSimplifier.scala:
```diff
@@ -5,7 +5,7 @@ import scala.collection.immutable.{SortedMap, SortedSet}
 
 import mlscript.utils.*, shorthands.*
-import utils.TraceLogger
+import hkmc2.utils.TraceLogger
```
```suggestion
import hkmc2.utils.*
```

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala:
```diff
@@ -167,14 +169,35 @@ object ParseRule:
     ,
     Kw(`if`):
-      ParseRule("`if` keyword")(
+      ParseRule("`if`")(
```
Why this change?

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala:
```diff
@@ -167,14 +169,35 @@ object ParseRule:
+      ParseRule("`if`")(
         Expr(
-          ParseRule("`if` expression"):
+          ParseRule("`if` split")(
```
This is for error messages. Why introduce internal jargon (if "split") here?

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala:
```diff
@@ -73,13 +82,47 @@ enum Tree extends AutoLocated:
+    case BoolLit(value) => value.toString
+    case Block(stmts) => stmts.map(_.describe).mkString("{ ", "; ", " }")
+    case Let(lhs, rhs, body) => s"let $lhs = $rhs in $body"
+    case TermDef(k, symName, alphaName, rhs) =>
+      val name = symName.orElse(alphaName).fold("")(n => s" ${n.describe}")
+      val sign = rhs.fold("")(s => s" : ${s.describe}")
+      s"${k.str}$name$sign"
```
This is crazy. I assume this is Copilot of ChatGPT junk. The goal of `describe` is simply to name what tree we're looking at using a word or two.

chengluyu replied:
You're absolutely right. I didn't look into this case. I will remove it soon.

chengluyu committed:
Update hkmc2/shared/src/main/scala/hkmc2/bbml/TypeSimplifier.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Implement `describe` in the right way

chengluyu committed:
Provide meaningful parsing rule descriptions

chengluyu committed:
Merge remote-tracking branch 'lptk/hkmc2' into hkmc2-ucs

# Conflicts:
#	hkmc2/jvm/src/test/scala/hkmc2/MLsDiffMaker.scala
#	hkmc2/shared/src/test/mlscript/bbml/bbBasics.mls
#	hkmc2/shared/src/test/mlscript/bbml/bbCheck.mls
#	hkmc2/shared/src/test/mlscript/bbml/bbGPCE.mls
#	hkmc2/shared/src/test/mlscript/bbml/bbQQ.mls
#	hkmc2/shared/src/test/mlscript/meta/ImporterTest.mls

chengluyu committed:
Show `Nil` in debug trees

chengluyu committed:
Add `ctx` command which shows new members and locals

chengluyu committed:
Support symbolic names for classes

chengluyu committed:
Support operator splits

chengluyu committed:
Mark extractors as `infix` to remove warnings

chengluyu committed:
Migrate the binary search tree example

Also support:
1. Parsing operator splits correctly
2. Elaborating on operator splits
3. Defining auxiliary functions via `let f(x) = ...`
4. Using blocks as function body

chengluyu committed:
Add the first example of operator splits in the paper

chengluyu committed:
Simplify the display of `FldFlags`

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Desugarer.scala:
```diff
@@ -5,31 +5,32 @@ import syntax.{Keyword, Tree}, Tree.*
 import mlscript.utils.*, shorthands.*
 import Message.MessageContext
 import utils.TraceLogger
+import scala.annotation.showAsInfix
```
Unused import...

LPTK commented on hkmc2/shared/src/test/mlscript/ucs/syntax/SimpleUCS.mls:
```diff
@@ -93,20 +84,14 @@ fun f(x, y, z) =
-:fixme
-// :e
+
+// :e // Note: should report missing final else clause.
```
I guess this should be both `:e` and `:fixme`, which should not be reported as a test failure.

chengluyu replied:
It is reported unless we disable the following lines.

https://github.com/LPTK/mlscript/blob/20fa71a262dcf659993172f9c6cfaafb998b77a6/hkmc2/jvm/src/test/scala/hkmc2/DiffMaker.scala#L167-L169

LPTK commented on hkmc2/shared/src/test/mlscript/ucs/syntax/Split.mls:
```diff
@@ -0,0 +1,130 @@
+    A then "A"
+//│ /!!!\ Uncaught error: scala.NotImplementedError: List()
+
+x
+  + 1
```
This is misparsed. Clearly it should not parse the same as `x({ +(1) })`. You can see `x` is not being applied and `+`  does not have a single argument...

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala:
```diff
@@ -280,8 +339,13 @@ class Elaborator(tl: TraceLogger)(using raise: Raise, state: State):
           case id: Ident =>
             (id, Nil, N, ctx)
+          case InfixApp(lhs, Keyword.`:`, rhs) =>
+            // For self-annotations
```
What are self-annotations?

LPTK committed:
Minor misc. improvements to UCS PR

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Split.scala:
```diff
@@ -3,70 +3,103 @@ package semantics
 
 import mlscript.utils.*, shorthands.*
 import syntax.*
+import scala.annotation.tailrec
```
Unused import

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/ParseRule.scala:
```diff
@@ -249,6 +306,19 @@ object ParseRule:
+    //     )):
+    //       case (consequent, N) => (test: Tree) => InfixApp(test, `then`, consequent)
+    //       case (consequent, S(default)) => (test: Tree) =>
+    //         InfixApp(test, `then`, InfixApp(consequent, `else`, default)),
```
Is this commented code still relevant? Or should it be removed?

LPTK reviewed (changes requested):
Here are some comments. I'll still merge the PR to avoid blocking progress, but I'll leave the comments open and you should make sure to address them later on, when you have time. I'll make the merge after you address the serious problem I told you about.

LPTK committed:
Remove warnings from compiler project

chengluyu committed:
Remove useless imports

Mentioned by @LPTK in https://github.com/LPTK/mlscript/pull/10

chengluyu committed:
Fix a misnaming

Mentioned by @LPTK in https://github.com/LPTK/mlscript/pull/10

chengluyu committed:
Remove useless comments

Mentioned by @LPTK in https://github.com/LPTK/mlscript/pull/10

chengluyu committed:
Fix a misparsed test case

Mentioned by @LPTK in https://github.com/LPTK/mlscript/pull/10#discussion_r1773430880

chengluyu committed:
Suppress the lack of errors with `todo`

and support commands ending with comments.

This is to address https://github.com/LPTK/mlscript/pull/10#discussion_r1773426856

chengluyu committed:
Add `OpBlock` for cont'd expressions & operator splits

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/Parser.scala:
```diff
@@ -547,27 +548,33 @@ abstract class Parser(
-    case (_, loc) :: _ =>
-      err(msg"Expect an operator" -> S(loc) :: Nil)
-      errExpr :: Nil
+  /** Parse a operator block. Each block item should be a binary operator
```
```suggestion
  /** Parse an operator block. Each block item should be a binary operator
```

chengluyu committed:
Fix ill-formed let bindings made by the desugarer

chengluyu committed:
Fix a grammatical mistake in a comment

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Migrate and fix a test about nested op splits

chengluyu committed:
Migrate two tree tests and stop making signatures

chengluyu committed:
Remove the use of `indent` to support JDK 7/8

chengluyu committed:
Remove the use of `isBlank` to support JDK 7/8
